### PR TITLE
Updated the hostname resource to remove WMI support and use PowerShell…

### PR DESCRIPTION
**Pull request by johnmccrae**
… to change Windows device names. Functional tests were added too.

Signed-off-by: John McCrae <jmccrae@chf.io>

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (non-breaking change that does not add functionality or fix an issue)

## Checklist:

- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have run the pre-merge tests locally and they pass.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] All commits have been signed-off for [the Developer Certificate of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

### CODE REVIEW

File: lib/chef/resource/hostname.rb [hunk 1]
```diff
@@ -243,13 +272,26 @@ def updated_ec2_config_xml
...
 
           unless Socket.gethostbyname(Socket.gethostname).first == new_resource.hostname
-            converge_by "set hostname to #{new_resource.hostname}" do
-              powershell_exec! <<~EOH
-                $sysInfo = Get-WmiObject -Class Win32_ComputerSystem
-                $sysInfo.Rename("#{new_resource.hostname}")
-              EOH
+            if is_domain_joined?
+              if new_resource.domain_user.nil? || new_resource.domain_password.nil?
+                raise "Domain Username and Password are required parameters"
```
Comment (tas50) [inline lib/chef/resource/hostname.rb:277]:
```suggestion
                raise "The `domain_user` and `domain_password` properties are required to change the hostname of a domain-connected Windows system."
```

Comment (johnmccrae) [reply lib/chef/resource/hostname.rb]: corrected

File: lib/chef/resource/hostname.rb [hunk 2]
```diff
@@ -243,13 +272,26 @@ def updated_ec2_config_xml
...
-                $sysInfo.Rename("#{new_resource.hostname}")
-              EOH
+            if is_domain_joined?
+              if new_resource.domain_user.nil? || new_resource.domain_password.nil?
+                raise "The `domain_user` and `domain_password` properties are required to change the hostname of a domain-connected Windows system."
+              else
+                converge_by "set hostname to #{new_resource.hostname}" do
+                  powershell_exec! <<~EOH
+                    $user = #{new_resource.domain_user}
+                    $secure_password = #{new_resource.domain_password} | Convertto-SecureString -AsPlainText -Force
```
Comment (sam1el) [inline lib/chef/resource/hostname.rb:282]: I would make the suggestion that in any place that the password can be seen, even if securestring, you may still want to ass the `sensitive true` to the resource block. In the field we find that customers still aren't okay with the obfuscated password being show.

File: lib/chef/resource/hostname.rb [hunk 3]
```diff
@@ -103,7 +128,11 @@ def updated_ec2_config_xml
         end
       end
 
-      action :set, description: "Sets the node's hostname." do
+      def is_domain_joined?
+        powershell_exec!("(Get-WmiObject -Class Win32_ComputerSystem).PartofDomain").result
+      end
+
+      action :set, description: "Sets the node's hostname" do
```
Comment (tas50) [inline lib/chef/resource/hostname.rb:135]:
```suggestion
      action :set, description: "Sets the node's hostname." do
```

File: lib/chef/resource/hostname.rb [hunk 4]
```diff
@@ -243,13 +272,26 @@ def updated_ec2_config_xml
...
+                    $user = #{new_resource.domain_user}
+                    $secure_password = #{new_resource.domain_password} | Convertto-SecureString -AsPlainText -Force
+                    $Credentials = New-Object System.Management.Automation.PSCredential -Argumentlist ($user, $secure_password)
+                    Rename-Computer -NewName #{new_resource.hostname} -DomainCredential $Credentials
+                  EOH
+                end
+              end
+            else
+              converge_by "set hostname to #{new_resource.hostname}" do
+                powershell_exec! <<~EOH
```
Comment (tas50) [inline lib/chef/resource/hostname.rb:290]: You can do this all on a single line

**Review by johnmccrae — COMMENTED** [review]
All windows tests are passing. I included/updated today's feedback.

File: lib/chef/resource/hostname.rb [hunk 5]
```diff
@@ -103,6 +128,10 @@ def updated_ec2_config_xml
         end
       end
 
+      def is_domain_joined?
+        powershell_exec!("(Get-WmiObject -Class Win32_ComputerSystem).PartofDomain").result
```
Comment (mwrock) [inline lib/chef/resource/hostname.rb:132]: should use `Get-CimInstance` over `Get-WmiObject`

**Review by mwrock — CHANGES_REQUESTED** [review]

File: lib/chef/resource/hostname.rb (hunk 5 above)
Comment (johnmccrae) [reply lib/chef/resource/hostname.rb:132]: thanks. Adding that now

**Review by mwrock — APPROVED** [review]

**Pull request merged**